crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -75,24 +75,35 @@ def current_step_state
       progress[:steps][current_step] || {}
     end
 
-    def start_step
+    # 'step' is name of the step user wants to start.
+    def start_step(step_name)
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        if running?
+        if running? step_name
           @logger.warn("The step has already been started")
           return false
         end
-        progress[:steps][current_step][:status] = :running
+        unless step_allowed? step_name
+          @logger.error("The start of step #{step_name} is requested in the wrong order")
+          return false
+        end
+        progress[:current_step] = step_name
+        progress[:steps][step_name][:status] = :running
         save
       end
     end
 
-    def end_step(success = true, errors = {})
+    def end_step(step_name = nil, success = true, errors = {})
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        unless running?
+        step_name ||= current_step
+        unless running? step_name
           @logger.warn("The step is not running, could not be finished")
           return false
         end
-        progress[:steps][current_step] = {
+        unless step_allowed? step_name
```

Comment:
Hm, true, it's not needed any more, since at this point, there's only `current_step`. Speaking of which... passing step name to `end_step` is no longer needed with this state of API...